src/linux/file_operations_linux.cpp

```diff
@@ -5,201 +5,255 @@
+LinuxFileOperations::LinuxFileOperations() : fd_(-1) {
+}
 
 LinuxFileOperations::~LinuxFileOperations() {
-  Close();
+    Close();
 }
 
-FileError LinuxFileOperations::OpenDevice(const std::string& path) {
-  return OpenInternal(path.c_str(), O_RDWR | O_SYNC);
+// Map a few common errno values to cross-platform details
+DetailedError LinuxFileOperations::MapErrnoDetail(int e) {
+    switch (e) {
+    case ENOSPC:
+        return DetailedError::kNoSpace;             // no space left
+    case EROFS:
+        return DetailedError::kWriteProtected;      // read-only FS
+    case EBUSY:
+        return DetailedError::kBusy;                // device/file busy
+    case EACCES:
+    case EPERM:
+        return DetailedError::kAccessDenied;        // permissions
+    case EINVAL:
+        return DetailedError::kInvalidParameter;    // bad arg/offset
+    case EIO:
+    case ENXIO:
+        return DetailedError::kIoDevice;            // I/O error
+    default:
+        return DetailedError::kNoDetails;
+    }
 }
 
-FileError LinuxFileOperations::CreateTestFile(const std::string& path, std::uint64_t size) {
-  // First create/truncate the file
-  FileError result = OpenInternal(path.c_str(), 
-                                  O_CREAT | O_RDWR | O_TRUNC, 
-                                  S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
-  if (result != FileError::kSuccess) {
-    return result;
-  }
+FileError LinuxFileOperations::OpenDevice(const std::string &path) {
+    return OpenInternal(path.c_str(), O_RDWR | O_SYNC);
+}
 
-  // Set the file size
-  if (ftruncate(fd_, static_cast<off_t>(size)) != 0) {
-    Close();
-    return FileError::kSizeError;
-  }
+FileError LinuxFileOperations::CreateTestFile(const std::string &path,
+                                              std::uint64_t size) {
+    // First create/truncate the file
+    FileError result = OpenInternal(path.c_str(), O_CREAT | O_RDWR | O_TRUNC,
+                                    S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
+    if (result != FileError::kSuccess) {
+        return result;
+    }
+
+    // Set the file size
+    if (ftruncate(fd_, static_cast<off_t>(size)) != 0) {
+        SetLastErrFromErrno(FileError::kSizeError);
+        Close();
+        return FileError::kSizeError;
+    }
 
-  return FileError::kSuccess;
+    ClearLastErr();
+    return FileError::kSuccess;
 }
 
-FileError LinuxFileOperations::WriteAtOffset(
-    std::uint64_t offset,
-    const std::uint8_t* data,
-    std::size_t size) {
-  
-  if (!IsOpen()) {
-    return FileError::kOpenError;
-  }
+FileError LinuxFileOperations::WriteAtOffset(std::uint64_t offset,
+                                             const std::uint8_t *data,
+                                             std::size_t size) {
 
-  if (lseek(fd_, static_cast<off_t>(offset), SEEK_SET) == -1) {
-    return FileError::kSeekError;
-  }
+    if (!IsOpen()) {
+        last_error_ = {FileError::kOpenError, DetailedError::kNoDetails, 0};
```

Comment:
Why isn't this using the SetLastErrFromErrno that's used elsewhere?

Reply:
Ah, fair enough. It just jumped out at me as a bit weird that this is modifying `last_error_` directly, whereas most other places seem to be accessing / modifying it via helper functions?